config/opal_check_package2.m4

```diff
@@ -0,0 +1,232 @@
+dnl -*- autoconf -*-
+dnl
+dnl Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
+dnl                         University Research and Technology
+dnl                         Corporation.  All rights reserved.
+dnl Copyright (c) 2004-2005 The University of Tennessee and The University
+dnl                         of Tennessee Research Foundation.  All rights
+dnl                         reserved.
+dnl Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+dnl                         University of Stuttgart.  All rights reserved.
+dnl Copyright (c) 2004-2005 The Regents of the University of California.
+dnl                         All rights reserved.
+dnl Copyright (c) 2012-2017 Cisco Systems, Inc.  All rights reserved.
+dnl Copyright (c) 2012      Oracle and/or its affiliates.  All rights reserved.
+dnl Copyright (c) 2014      Intel, Inc. All rights reserved.
+dnl Copyright (c) 2015-2018 Research Organization for Information Science
+dnl                         and Technology (RIST). All rights reserved.
+dnl $COPYRIGHT$
+dnl
+dnl Additional copyrights may follow
+dnl
+dnl $HEADER$
+dnl
+
+dnl OPAL_DECLARE_PACKAGE(name,[help],[cppflagshelp],[ldflagshelp])
+AC_DEFUN([OPAL_DECLARE_PACKAGE],[
+    AC_ARG_WITH([$1], [AC_HELP_STRING([--with-$1=DIR], [$2])])
+    AC_ARG_WITH([$1-cppflags], [AC_HELP_STRING([--with-$1-cppflags=FLAGS], [$3])])
+    AC_ARG_WITH([$1-ldflags], [AC_HELP_STRING([--with-$1-ldflags=FLAGS], [$4])])
+    AC_ARG_WITH([$1-libdir], [AC_HELP_STRING([--with-$1-libdir=DIR],
+                                             [Deprecated, use --with-$1-ldflags instead])])
+    AS_IF([test -n "$with_$1_libdir"],
+          [AS_IF([test -n "$with_$1_ldflags"],
+                 [AC_MSG_WARN([It is not possible to use both --with-$1-ldflags and --with-$1-libdir])
+                  AC_MSG_ERROR([Cannot continue])],
+                 [AC_MSG_WARN([--with_$1_libdir=DIR is deprecated, use --with-$1_ldflags=FLAGS instead])
+                  with_$1_ldflags="-L$with_$1_libdir"])])
+])
+
+dnl --------------------------------------------------------------------
+dnl _OPAL_CHECK_PACKAGE2_HEADER(package, prefix, header, includes,
+dnl                             [action-if-found], [action-if-not-found])
+dnl --------------------------------------------------------------------
+AC_DEFUN([_OPAL_CHECK_PACKAGE2_HEADER], [
+    AS_VAR_PUSHDEF([opal_With], [with_$1])
+    AS_VAR_PUSHDEF([opal_Cppflags], [with_$1_cppflags])
+
+    AS_IF([test -n "$opal_Cppflags"],
+          [$2_CPPFLAGS="$$2_CPPFLAGS $opal_Cppflags"
+           CPPFLAGS="$CPPFLAGS $opal_Cppflags"
+           AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+          [AS_IF([test -z "$opal_With" || test "$opal_With" = "yes"],
+                 [AC_CHECK_HEADERS([$3], [$5], [$6], [$4])],
+                 [AS_IF([test -r "$opal_With/$3"],
```

Comment:
We agreed we were getting rid of this, didn't we?  We were *only* going to check with_dir/include if with_cppflags wasn't set.  There should only need to be a single call to AC_CHECK_HEADERS in this function, with a simple set if if statements to set CPPFLAGS as needed.

Reply:
The current code does __not__ check `DIR` nor `DIR/include` if  `--with-FOO-cppflags` is set, did I get that part wrong ?
Also, the current code fails if `--with-FOO-cppflags` is not used __and__ nor `DIR/header.h` nor `DIR/include/header.h` exists. If we move to a single `AC_CHECK_HEADERS` invokation, handling this bozo case might be a bit convoluted. I am open to suggestions though.

Reply:
@ggouaillardet I think the use of the ```test -n``` option is what confused here - I honestly don't think we have used that variation elsewhere. Nothing wrong with it - just something I had to look up (and ask about), and still trip over until I get used to it.

The code looks generally correct to me. If they provide the cppflags option, then we just use it. If not, then we check std locations if opal_with is empty or "yes", then DIR, and then DIR/include.  The only thing I believe we might change is to reduce the complexity a bit.

I believe what @bwbarrett was getting at is that you could use all those AS_IF's to simply set a local cppflags value, and then set CPPFLAGS and have a single call to ```AC_CHECK_HEADERS``` at the end. Would significantly simplify the code.

Reply:
The test -n didn't confuse me.  I was saying that the code structure is way more complicated than it needs to be.  It should be something like:

```
AS_IF([test -n "$with_FOO_cppflags"], [CPPFLAGS=$with_FOO_cppflags],
  AS_IF([test -n "$with_FOO" != "yes"], [CPPFLAGS=$with_foo/include])])
AC_CHECK_HEADERS([....])
```

Reply:
@rhc54 @bwbarrett my question was how to elegantly handle the case in which `--with-FOO-cppflags` is not set __and__ `--with-FOO=DIR` is used __and__ nor `DIR/foo.h` nor `DIR/include/foo.h` exist.
Ideally, it would fail without aborting nor invoking `AC_CHECK_HEADERS`.

Reply:
Why wouldn't it abort? The user specified a DIR, and you didn't find any such header file. It seems to me that this must abort, though I'm okay with the "abort" coming from the caller once this returns the error if that is what you are suggesting. All we were suggesting is that the code can be simplified.

Reply:
I don't think that ensuring DIR/include/foo.h (and remember, we're *NOT* checking for DIR/foo.h) is important.  If you feel strongly that it is for the case of --with-dir, I'd do something like:

```
if test -n "$with_FOO_cppflags" ; then
   add_CPPFLAGS="$with_FOO_cppflags"
elif test "$with_FOO" != "yes"; then
    if test ! -r "$with_FOO/include/$header_file"; then
         AC_MSG_ERROR([Didn't find expected header <blah> in <blah>.])
    fi
   add_CPPFLAGS="-I$with_FOO/include"
fi
CPPFLAGS=$CPPFLAGS $add_CPPFLAGS
AC_CHECK_HEADERS
```

Again, this all got so complicated it had a billion unintended side effects.  Our whole goal is to get back to simple, even if we miss some corner cases in testing; we can define what we do, how it doesn't do what the user may want, but that the important part is *IT'S DEFINED*.

Reply:
Huh??? Part of this whole thing was to ensure that we _do_ check for DIR/foo.h, and then for DIR/include/foo.h, as that is where multiple vendors install by default. I don't understand why that is objectionable.

Reply:
No, that's terrible.  --with-FOO should *never* add DIR/foo.h.  The whole point of this change is to add --with-FOO-cppflags so that if the header was in an odd spot, customers could add the right CPPFLAGS to get what they need.  DIR/foo.h adds a really dangerous -I flag in all but some corner-case situations.

Reply:
I give up - this has morphed into something that doesn't appear to meet my needs at all. I'll just hardcode what I need into the .m4's where I need it.

Reply:
@bwbarett One motivation for the change is to allow users to install their include files in a non standard location (in this case, it was `/usr/include/pmix` instead of `/usr/include`), so we cannot assume we only need to append `/include` to `DIR` and look for headers there.

That being said, and since `--with-FOO-cppflags` and `--with-FOO-ldflags` are new options, the new macro will have to differ between `master` and the release branches for a while. As far as I am concerned, I would be fine testing for `DIR/foo.h` only in the release branches, and `master` not testing it (since `--with-FOO-cppflags` can be used to achieve the same goal).

@rhc54 is this acceptable for you ?